### DISCUSSION TITLE
How to migrate large .NET Framework based WPF application to Avalonia

### DISCUSSION THREAD

**Discussion by schandra09net** [AvaloniaUI/Avalonia#10777] [Q&A]
Hello Team,

**My existing application is as follows:**

1. C# .NET v4.5.2 WPF application (Dependent on multiple inhouse library projects developed with v4.5.2)
2. Has a lot of 3rd parties NuGet dependencies.
3. Has a lot of Win32 API calls.

**To port this application to Avalonia is my below understanding correct:**
1. First, port all in-house library projects from .Net 4.5.2 to >=.NET5
2. Then create a new Avalonia application
a. Recreate every view to Avalonia .axaml from .xaml from my old WPF application.
b. Replace/modify the not supported style, template, resource, etc. to the new way/implementation as per Avalonia.

**A couple of questions here:**
1. Can I directly use the existing WPF code with minimal changes or a converter/tool to the Avalonia?
2. How to replace Win32 APIs, will they work as it is after conversion?
**ex:** 
static extern bool SetWindowPos(params);
static extern IntPtr FindWindow(params); etc.
3. How to deal with NuGet packages that are not supported by .NET Standard or >=.NET5?
**ex:** 
https://www.nuget.org/packages/Metrics.NET
https://www.nuget.org/packages/Unity etc.

Any other things I should be aware of, please let me know.

**Comment by kyuranger** [comment]
You can use the Avalonia XPF to achieve this easily. However, it is not free.

Otherwise, you have to rewrite all of the code one by one (just like me).

**Comment by timunie** [comment]
1. read through this: https://docs.avaloniaui.net/misc/wpf
2. you can use our XPF, but it's not free
3. we have also paid support if needed
4. You can use .NET Framework with Avalonia
5. You can use WinForms or WPF dependencies, however it will only work on Windows.

**Reply by timunie** [reply]
Here is the link to the XPF request page if someone is interested: 
- What it is: https://avaloniaui.net/XPF
- Request a demo: https://avaloniaui.net/XPF/Interest

**Comment by schandra09net** [comment]
Hi Tim,

Thanks for your response, I am looking at porting our existing WPF application to be Cross-Platform(Windows, MacOS, and Linux)

1.  read through this: https://docs.avaloniaui.net/misc/wpf
**schandra:** _So, the changes should be done manually file by file adhering to the https://docs.avaloniaui.net/misc/wpf suggestions._

2. you can use our XPF, but it's not free
**schandra:** _How can I get this XPF (NuGet or SDK ...) and will Avalonia provide a trial version to explore before going to get the full version?_
_and is there a way to use an analyzer using XPF that will give more details before porting/migration from WPF?_

3. we have also paid support if needed
**schandra:** _Yes, I understand that the paid support after going through the Avalonia website._

4. You can use .NET Framework with Avalonia
5. You can use WinForms or WPF dependencies, however, it will only work on Windows.
**schandra:** _If I am looking to port my application to be Cross-Platform, I don't think the need of using .Net Framework with Avalonia will be useful. correct me if I am wrong._

**Reply by timunie** [reply]
2. -> Please reach out to the team via mail. They can analyze your project and make an offer. I'm sure there will be an option to test it before. 
3.
4.
5. -> In that case, you should go with .NET 6 or .NET 7 and also remove references to Windows. If you need platform specific code, it can be done via Services I think.
